Support shipping directly to monitoring cluster

ycombinator opened the pull request:
Currently all beats ship their x-pack monitoring data to the production Elasticsearch cluster by `POST`ing it to the custom `_xpack/monitoring/_bulk` Elasticsearch API endpoint.

In the future we want to support beats shipping their x-pack monitoring data directly to a monitoring Elasticsearch cluster (which _could_ be the same as their production cluster but doesn't necessarily have to be) by `POST`ing it to the regular `_bulk` Elasticsearch API endpoint.

This PR introduces this "direct shipping" feature. Concretely:

* [x] It introduces a new class of settings named `monitoring.*` that will parallel the current `xpack.monitoring.*` settings. If the user has set `xpack.monitoring.*` the beat will continue to ship to the production cluster as it does today (but also emit a deprecation warning about these settings). However, if the user has set the `monitoring.*`, the beat will instead ship directly to the monitoring cluster.

* [x] It modifies the Elastisearch reporter to decide where to ship the data (production cluster, `_xpack/monitoring/_bulk` API endpoint -or- monitoring cluster, `_bulk` API endpoint).

* [x] If data is being shipped directly to the monitoring cluster AND the beat is configured to use an `elasticsearch` output, the Elasticsearch reporter will inject this `cluster_uuid` of the `elasticsearch` output cluster into the monitoring data event before publishing it.

ycombinator commented:
@urso @ruflin This PR is very much a WIP but I'd like some early feedback from you to know if I'm on the right track (since this is my first non-trivial PR to libbeat code).

Specifically at this point I'm wondering what you think of how/when I'm getting the `cluster_uuid` from the ES output cluster and then how I can properly use it in the ES reporter to add it to the monitoring event.

Note that  I haven't yet added the logic to ship to the regular `_bulk` API on the monitoring cluster instead of the custom `_xpack/monitoring/_bulk` API on the production cluster.

urso commented on libbeat/cmd/instance/beat.go:
```diff
@@ -357,11 +359,16 @@ func (b *Beat) launch(settings Settings, bt beat.Creator) error {
+		monitoringCfg, err := chooseMonitoringConfig(b.Config)
+		if err != nil {
+			return err
+		}
```
how about:

```

if monitoringCfg, err := selectMonitoringConfig(b.Config); err != nil {
  return err
} else if monitoringCfg.Enabled() {
  ...
  reporter, err := report.New(b.Info, settings, monitoringCfg, b.Config.Output)
  ...
}
```

with `selectMonitoringConfig`:

```
func selectMonitoringConfig(cfg *common.Config) (*common.Config, error) {
  switch {
    case cfg.Monitoring.Enabled() && cfg.MonitoringNew.Enabled():
      return nil, errMonitoringBothEnabled
    case cfg.Monitoring.Enabled():
      ...
    case cfg.MonitoringNew.Enabled():
      ...
    default:
      return nil, nil
  }
}
```

urso commented on libbeat/cmd/instance/beat.go:
```diff
@@ -95,8 +96,9 @@ type beatConfig struct {
-	Monitoring *common.Config  `config:"xpack.monitoring"`
+	Pipeline      pipeline.Config `config:",inline"`
+	Monitoring    *common.Config  `config:"xpack.monitoring"`
+	MonitoringNew *common.Config  `config:"monitoring"`
```
Let's call them `XPackMonitoring` and `Monitoring`.

ycombinator replied:
Since these are exported field names, are you sure it's okay to "break" the API like that?

urso replied:
They are semi-exported :)

The beatConfig type is not exported, that is fields can not be accessed by other packages.
Yet the fields inside beatConfig are exported, so they can be correctly filled in via reflection when unpacking the config. The `config:"<name>"` determines the configuration key name. The `Monitoring` and `MonitoringNew` fields names are only for internal representation.

urso commented on libbeat/cmd/instance/beat.go:
```diff
@@ -783,6 +790,36 @@ func (b *Beat) templateLoadingCallback() (func(esClient *elasticsearch.Client) e
+		return nil
+	}
+
+	return callback, nil
```
The callback will be run on every reconnect. The first connect only occurs if the beat actually has an event to publish. If there is no event or the output is not available on start we can not fetch the cluster id at all. Will this be an issue in the UI?

If possible I'd prefer not to include the ES cluster ID.

ycombinator replied:
> ... or output is not available on start

Can you clarify what this means or how this scenario can occur?

> If possible I'd prefer not to include the ES cluster ID.

Include it in the monitoring events/docs? Or somewhere else?

urso replied:
>> ... or output is not available on start
> Can you clarify what this means or how this scenario can occur?

When beats startup, but is idle, then we do not attempt to connect to Elasticsearch. That is, the callback checking for the cluster ID is never run. The callback will only applied if we have events to publish AND we sucessfully managed to connect and ping the remote ES host. If connection fails and event is eventually dropped (in case of metricbeat/packetbeat), then the callback will still not be executed.

Due to this we have no cluster ID, yet metrics are/must be published to the monitoring cluster all the time.

>> If possible I'd prefer not to include the ES cluster ID.
> Include it in the monitoring events/docs? Or somewhere else?

Not send the ES cluster ID at all. There is no guarantee if/when you have the cluster ID available. E.g. if the beat drops all events because the host setting is wrong, then we still want to get the monitoring metrics.
If the beat is configured to ship via Logstash/Kafka/Redis, then you won't have a cluster ID.

ycombinator replied:
I understand better now, thanks for the explanation.

> ...
> Due to this we have no cluster ID, yet metrics are/must be published to the monitoring cluster all the time.

Technically it should be okay to send monitoring docs without a `cluster_uuid`. The Monitoring UI will be able to handle that once https://github.com/elastic/kibana/issues/27595 is resolved.

> ...
> If the beat is configured to ship via Logstash/Kafka/Redis, then you won't have a cluster ID.

Good point. In this case we could ask users to configure `monitoring.elasticsearch.*`. 

--

Bottomline is that the `cluster_uuid` corresponds to the ES cluster that the beat is sending its data to via `output.elasticsearch.*`. If this output is not configured or we are not able to connect to that cluster yet or ever, there's no way we can figure out its `cluster_uuid`. So there's no way we can, or even should, try to associate monitoring data with this cluster.

urso replied:
right.

In the end it’s a correlation problem. It would be easier for us if we would send the/one hostname and the monitoring endpoint will do the lookup in order to fill the cluster id.

> Good point. In this case we could ask users to configure monitoring.elasticsearch.*

makes sense. But this Elasticsearch might not be reachable. For this case it we could also introduce a `montoring.elasticsearch.cluster_id` setting.

ycombinator replied:
We had this exact discussion amongst the stack monitoring team. In fact, the original plan was to include exactly a setting like you mentioned. The concern was that it puts adds to the user's burden and, after discussion, we decided a better solution was to make it so that the Monitoring UI could handle monitoring data with no `cluster_uuid` in it. So we are going down that path now: https://github.com/elastic/kibana/pull/28278.

urso commented on libbeat/cmd/instance/beat.go:
```diff
@@ -886,3 +923,17 @@ func obfuscateConfigOpts() []ucfg.Option {
+		monitoringCfg = beatCfg.MonitoringNew
+		monitoringCfg.SetString("format", -1, "monitoring")
+	}
+	return
```
style:
- do not use fmt.Errorf with a constant string. Prefer a global variable initialized with `errors.New`.
- for longer if-else cascade you can use `switch { }`
- it's ok to use named return values for documentation purposes, but better avoid naked returns all together. ([This example](https://play.golang.org/p/6D3tIYLBz0T) displays the problem with naked returns and shadowing).

urso commented on libbeat/monitoring/report/elasticsearch/client.go:
```diff
@@ -113,22 +115,15 @@ func (c *publishClient) Publish(batch publisher.Batch) error {
+			err = c.bulkToMonitoring(event)
+		default:
+			err = fmt.Errorf("unsupported reporting format: %v", params["format"])
 		}
```
I always feel a little uneasy about comparison on constant strings spread out in the code. As the format doesn't change I'd prefer to have an enum or bool flag which must be set when calling the constructor. 

The error message should be returned on init, and not when publishing is ongoing.  => Better have constructor check for it.

urso commented on libbeat/cmd/instance/beat.go:
```diff
@@ -886,3 +923,17 @@ func obfuscateConfigOpts() []ucfg.Option {
+		monitoringCfg.SetString("format", -1, "production")
+	} else {
+		monitoringCfg = beatCfg.MonitoringNew
+		monitoringCfg.SetString("format", -1, "monitoring")
```
Are users allowed to set `format` themselves? Maybe prefix with an underscore.

urso commented on libbeat/monitoring/report/elasticsearch/client.go:
```diff
@@ -150,3 +145,45 @@ func (c *publishClient) Test(d testing.Driver) {
+func (c *publishClient) bulkToMonitoring(event publisher.Event) error {
+	action := common.MapStr{
+		"index": common.MapStr{
+			"_type":    "doc",
```
this one is annoying. If the target ES version is 7.0 we will have to remove the `_type` field.

urso commented on libbeat/monitoring/report/report.go:
```diff
@@ -110,6 +110,12 @@ func getReporterConfig(
+		if err != nil {
+			return "", nil, err
+		}
+		rc.SetString("parameters.format", -1, format)
```
uff. I wish we could make this 'cleaner'.

Why have this config check here? The evaluation of the `format` setting is better handled in the elasticsearch reporter, as it's a local 'feature' to the ES reporter only.

ycombinator replied:
Yeah, I'm not a big fan of this either, but the format depends on whether the user is setting `xpack.monitoring.*` or `monitoring.*`. This code here coupled with https://github.com/elastic/beats/pull/9260/files#diff-f7226280e7734d6e3d88e7dde232602eR1010 is letting us determine the format based on the configuration and passing it down to the reporter(s).

I couldn't come up with a nicer way to pass down to the ES reporter whether the user had set `xpack.monitoring.*` or `monitoring.*`. Any ideas?

urso replied:
As we dynamically build a config object here, we can't do that much. We still need to forward the `_format` setting somehow. This can be done by either:
1. extract the configuration and pass it as extra parameter to the reporters constructor
2. write a setting in the constructed configuration object

You did opt for solution number 2. None is really nice/clean here, so I'm fine with this chosen approach. What feels somewhat more hacky about it is that you overwrite/add to the `parameters` setting, which is already a setting of the elasticsearch reporter, used to set additional HTTP parameters. I think I would prefer the reporter itself to interpret `_format` the way it is needed. That is: replace `rc.SetString("parameters._format", -1, _format)` with `rc.SetString("_format", -1, _format)`, and have the reporter read and apply the `_format` setting as required. E.g. have the ES reporter read the `_format` setting and add it internally to it's parameters. This way it's always the same setting we're dealing with, pushing the semantic (how to use it) to the reporter, not anywhere inbetween. This way `_format` becomes less intermingled with user-provided settings.

ycombinator replied:
Oh, it wasn't obvious to me that the `parameters` setting was meant just for HTTP parameters, but I see that now. Thanks for drawing my attention to this.

I agree then that we shouldn't corrupt that setting like I am doing at the moment. Will fix per your recommendation.

ycombinator commented:
@urso @ruflin I've just finished some of the pending work on this PR and it is ready for your review again. Thanks!

ruflin commented:
Is this already ready to go? Asking because I see https://github.com/elastic/kibana/issues/27595 is still open?

If someone configures `monitoring.*` with the same values as in `x-pack.monitoring` the outcome will be identical?

What is the expected behaviour if someone connects to an OSS cluster? Will the data be shipped or not?

ycombinator commented:
> Is this already ready to go? Asking because I see elastic/kibana#27595 is still open?

This is ready for **review**. But, as the note at the top of this PR says, it should not be merged until elastic/kibana#27595 is merged first. Basically I would like to have this PR reviewed and ready to merge so we can just push the button as soon as elastic/kibana#27595 gets merged.

> If someone configures `monitoring.*` with the same values as in `x-pack.monitoring` the outcome will be identical?

Yes, the end result will be the same: exactly the same documents will end up in `.monitoring-beats-*` indices. The difference will be in _how_ they get there. If the user configures `xpack.monitoring.*`, they will get there via the `POST _xpack/monitoring/bulk` API on the **production ES cluster**. However, if the user configures `monitoring.*`, they will get there via the `POST _bulk` API on the **monitoring** cluster.

> What is the expected behaviour if someone connects to an OSS cluster? Will the data be shipped or not?

The data should not be shipped. I would expect the checks in this method to fail and corresponding errors to be logged:

https://github.com/elastic/beats/blob/4f826eda5fce8acf9868f3f164eb5b382909b5c6/libbeat/monitoring/report/elasticsearch/client.go#L48-L81

ruflin commented on libbeat/cmd/instance/beat.go:
```diff
@@ -362,11 +369,13 @@ func (b *Beat) launch(settings Settings, bt beat.Creator) error {
-	if b.Config.Monitoring.Enabled() {
+	if monitoringCfg, err := selectMonitoringConfig(b.Config); err != nil {
+		return err
+	} else if monitoringCfg.Enabled() {
```
Do you need an else here as you return above if it matches?

ruflin commented on libbeat/monitoring/report/elasticsearch/client.go:
```diff
@@ -77,6 +80,8 @@ func (c *publishClient) Connect() error {
 	}
 
 	debugf("XPack monitoring is enabled")
+	debugf("Publishing to %v cluster", c.params["_format"])
```
Do we know for sure `_format` is set? Want to make sure that because of log nothing is panicing.

ycombinator replied:
It should be set for sure at this point, but that also requires a check in the constructor. I had this at one point but alas it seems to have got lost in a rebase 😢. Will re-add.

ruflin commented on libbeat/monitoring/report/elasticsearch/elasticsearch.go:
```diff
@@ -358,3 +368,19 @@ func makeMeta(beat beat.Info) common.MapStr {
+	}
+
+	snapshot := monitoring.CollectFlatSnapshot(elasticsearchRegistry, monitoring.Full, false)
+	return snapshot.Strings["cluster_uuid"]
```
Is `cluster_uuid` always set?

ycombinator replied:
No, this is not guaranteed. It will be never set if the beat is not using the `elasticsearch` output OR if it is using the `elasticsearch` output but the initial connection has not yet been established.

Is your concern that `snapshot.Strings["cluster_uuid"]` might fail somehow? If so, this should not happen as `snapshot.Strings` is a `map[string]string` and if the map key `cluster_uuid` doesn't exist, the zero value of `string`, that is `""`, will be returned here.

ruflin replied:
yes, my concern was that it might panic if you try to access a key that not exist. If that is no tthe case, all fine.

ruflin commented on libbeat/monitoring/report/report.go:
```diff
@@ -25,6 +25,11 @@ import (
 )
 
+const (
+	ReportingFormatProduction = "production"
```
I stumbled over this `production` name several times during the review. I understand that the naming comes from shipping to the "Production" cluster. At the same time a monitoring cluster is also a production cluster. Perhaps we can name this internally `bulk` and `xpack` as the reporting type to make the difference more obvious?

ycombinator replied:
Ahhhh naming 😄 

> Perhaps we can name this internally `bulk` and `xpack` as the reporting type to make the difference more obvious?

I went with the `production` vs. `monitoring` cluster terminology as it's the same terminology we use in stack monitoring discussions so it's a vocabulary that's more "obvious" to most of us on the team (even if it might be flawed). Seeing these terms in debug messages and such when troubleshooting issues would instantly make sense to us IMO.

> At the same time a monitoring cluster is also a production cluster. 

Indeed, this is a fair point. Perhaps `monitoring` vs. `data` cluster then? Again the same problem arises: that the monitoring cluster also stores data, after all 😄. Also, this terminology doesn't align with the terminology being used in stack monitoring discussions but it does remove the use of `production` (as the monitoring cluster is also a production cluster).

Naming is hard and so I'm open to changing the naming here if reviewers feel particularly strongly in any one direction.

ycombinator replied:
> Perhaps we can name this internally `bulk` and `xpack` as the reporting type to make the difference more obvious?

BTW, the xpack endpoint also has `bulk` in it, so seeing `bulk` in the code or in a debug message could be a bit ambiguous too and might require some digging.

How about we be really explicit about the names of these constants (and not care so much about their values) then? Something like:

```go
type ReportingFormat int

const (
  ReportingFormatXPackMonitoringBulk ReportingFormat = iota
  ReportingFormatBulk
)
```

ruflin replied:
No strong feelings in any direction, just thought it's worth mentioning. I like your `iota` proposal.

ycombinator commented:
I started writing a system test for this "direct monitoring" feature per @ruflin's recommendation. While doing so I realized we don't have a system test for the current "indirect monitoring" feature either. So I'm going to first write that in https://github.com/elastic/beats/pull/10645.

ycombinator commented:
@urso @ruflin I've added tests and a CHANGELOG entry. This PR is ready for review again. Thanks!

ruflin commented on libbeat/tests/system/test_monitoring.py:
```diff
@@ -45,10 +47,109 @@ def test_via_output_cluster(self):
+
+        proc.check_kill_and_wait()
+
+        # TODO: Get a beats_stats doc
```
? leftovers or should stay?

ycombinator replied:
Worse, they actually need to be implemented in this PR. My mistake 🤦‍♂️. Implementing now...

ycombinator replied:
TODOs have been implemented now. Sorry about that!

ycombinator commented:
Alright @ruflin @urso Latest round of feedback addressed. Ready for review again. Thanks!

ycombinator commented on libbeat/cmd/instance/beat.go:
```diff
@@ -772,6 +784,47 @@ func (b *Beat) createOutput(stats outputs.Observer, cfg common.ConfigNamespace)
+		if err != nil {
+			return errw.Wrap(err, "error querying /")
+		}
+		if status > 299 {
```
Good question. Short answer is **yes**, redirects are handled by `esClient`.

Slightly longer answer is that `esClient` ultimately ends up calling https://golang.org/pkg/net/http/#Client.Do. By default, this will redirect up to 10 times before bailing out with a `stopped after 10 redirects` message.

urso commented on libbeat/cmd/instance/beat.go:
```diff
@@ -772,6 +784,47 @@ func (b *Beat) createOutput(stats outputs.Observer, cfg common.ConfigNamespace)
+		monitoring.NewString(elasticsearchRegistry, "cluster_uuid").Set(response.ClusterUUID)
+
+		return nil
+	}
```
If the user configures multiple hosts, this callback will be called per host, potentially at the same time. The use of `response` is not thread-safe (even if they report the same value, race detector will complain).

Plus each call will update the registry. Not sure right now, but NewString or NewRegistry might even panic if the "key" is already known.

urso commented on libbeat/cmd/instance/beat.go:
```diff
@@ -908,3 +961,23 @@ func initKibanaConfig(beatConfig beatConfig) (*common.Config, error) {
+	default:
+		return nil, nil
+	}
+}
```
Better write a detailed log message, but return a short/descriptive error value. Also think how you error message will print. Most likely: `Failed: both .... cannot be set. Prefer ... cluster host`, not creating a full sentence.

Why do: `const msg = ...;  return nil, errors.New(msg)`? No need to allocate an error. If you have a descriptive name for your error, then declare the error once: `var errMonitoringBothEnabled = errors.New(...)`

btw. can we abstract and move the configuration/loading part from the `instance` package to the `monitoring` (or some other) package?

ycombinator replied:
> Better write a detailed log message, but return a short/descriptive error value. Also think how you error message will print. Most likely: Failed: both .... cannot be set. Prefer ... cluster host, not creating a full sentence.

Apologies but I didn't quite understand the problem here. Were you suggesting this `selectMonitoringConfig` function should itself do logging but return a short/descriptive error type? Or something different?

> btw. can we abstract and move the configuration/loading part from the instance package to the monitoring (or some other) package?

I moved the `instance.selectMonitoringConfig` function to `monitoring.SelectConfig`. I had to make a few other changes as well, notably around the `instance.beatConfig` struct. I think it would be good if you could review these changes.

urso replied:
Phew... don't remember full context here and I can not find the code pattern or line with recent changes anymore.

Still I see a few `errors.New(<message>)`. This allocates a new error everytime you want to report an error. Errors returned this way can be logged, but are not actionable. One might consider to declare at least a sentinal error constant in the package like `var ErrShortName = errors.New(<common message>)` and reuse the constant.

ycombinator replied:
I'm still not 100% sure I got this right but please take a look at the changes I made in 56383a3ae.

urso commented on libbeat/monitoring/report/elasticsearch/client.go:
```diff
@@ -20,6 +20,9 @@ package elasticsearch
 	"fmt"
+	"time"
+
+	"github.com/pkg/errors"
```
did you mean to import `errors` or `pkg/errors`?

ycombinator replied:
`pkg/errors` as I'm using it for `errors.Wrap`.

urso commented on libbeat/monitoring/report/elasticsearch/client.go:
```diff
@@ -113,24 +120,13 @@ func (c *publishClient) Publish(batch publisher.Batch) error {
+		case report.ReportingFormatXPackMonitoringBulk:
+			err = c.publishWithXPackMonitoringBulk(params, event, t)
+		case report.ReportingFormatBulk:
+			err = c.publishWithBulk(event)
```
Could we find shorter but descriptive names? e.g. `publishBulk` and `publishXpackBulk` ?

urso commented on libbeat/monitoring/report/elasticsearch/client.go:
```diff
@@ -152,3 +148,88 @@ func (c *publishClient) Test(d testing.Driver) {
 	return "publish(" + c.es.String() + ")"
 }
+
+func (c *publishClient) publishWithXPackMonitoringBulk(params map[string]string, event publisher.Event, t interface{}) error {
```
I'd feel a little more comfortable we `t` would be a string. nit: naming (e.g. `doctype` or `typ`)

urso commented on libbeat/monitoring/report/elasticsearch/client.go:
```diff
@@ -152,3 +148,88 @@ func (c *publishClient) Test(d testing.Driver) {
+	v7 := common.MustNewVersion("7.0.0")
+	if esVersion.LessThan(v7) {
+		action.Put("index._type", "doc")
+	}
```
Create global const for v7 or check for esVersion.Major. Instead of `relying` on `action.Put` how about splitting it into action and meta, with `meta := common.MapStr{"_index", ..., "_routing"}`.

urso commented on libbeat/monitoring/report/elasticsearch/client.go:
```diff
@@ -152,3 +148,88 @@ func (c *publishClient) Test(d testing.Driver) {
+	fields := common.MapStr{
+		"type": tStr,
+		tStr:   event.Content.Fields,
+	}
```
We extract `t` and pass it when calling the `xpack` publisher method, but here we duplicate logic for extracting the type?

urso commented on libbeat/monitoring/report/report.go:
```diff
@@ -25,6 +25,16 @@ import (
+	ReportingFormatUnknown ReportingFormat = iota // to protect against zero-value errors
+	ReportingFormatXPackMonitoringBulk
+	ReportingFormatBulk
+)
```
This one is very specific to the Elasticsearch reporter only. Can we hide this detail in the ES reporter plugin only?

nit: Always consider the package name to be part of the full name of a symbol. How about: `type EndpointType int`  enumeration with `UnknownEndpointType`, `XpackMonitoringEndpoint`, and `BulkEndpoint`?

ycombinator replied:
I tried to do push these constants down into the `monitoring/report/elasticsearch` package  but then I end up with import cycles, since these constants are used in the `monitoring/report` package, for e.g. [here](https://github.com/elastic/beats/pull/9260/files#diff-417858b4623eafaefe6fd1504259a6a3R99) and [here](https://github.com/elastic/beats/pull/9260/files#diff-417858b4623eafaefe6fd1504259a6a3R115).

This was why I kept these constants in the `monitoring/report` package and named them something more generic like "format" instead of "endpoint". The latter is definitely specific to Elasticsearch. By naming it "format", my thinking was that other monitoring reporters in the future could choose to handle the different reporting formats (or not).

urso replied:
Well, alternative reporters are: write to log file, make available via HTTP REST API, and so on. No one besides the ES report is using any of these.
If we really need them to hack around limitation or how config is passed vs. to be interpreted, then it's fine for now. But in this case we should add a code comment explaining why it's here and if/how we plan to clean this up.

ycombinator replied:
As we discussed off-PR, I'm not happy about the hackiness of passing down this reporting "format", which is really only useful to the Elasticsearch reporter.

I couldn't come up with a really good solution without majorly rewriting how reporters configure themselves. However, I did a bit of cleanup and removed _some_ of the hackiness in 40e5163d976fe297ad00dc5f45a984d807b44125. Please let me know what you think of this solution.

ycombinator commented on libbeat/monitoring/monitoring.go:
```diff
@@ -17,7 +17,19 @@
+)
+
+// MonitoringBeatConfig represents the part of the $BEAT.yml to do with monitoring settings
+type MonitoringBeatConfig struct {
```
I originally tried to name this just `BeatConfig` to avoid the naming stutter. However, that didn't quite work out with composing this struct into the [`instance.beatConfig` struct](https://github.com/elastic/beats/pull/9260/files#diff-f7226280e7734d6e3d88e7dde232602eR87), since the latter already composes in the [`beat.BeatConfig` struct](https://github.com/elastic/beats/pull/9260/files#diff-f7226280e7734d6e3d88e7dde232602eL88), thereby resulting in a naming collision.

urso replied:
The naming conflict is due to:

```
	// monitoring settings
	monitoring.MonitoringBeatConfig `config:",inline"`
```

One could have resolved it via:

```
	// monitoring settings
	MonitoringBeatConfig monitoring.BeatConfig `config:",inline"`
```

When embedding a struct you create an anonymous field that is accessed via the struct it's name.

ycombinator replied:
Addressed in 4573a30.

ycombinator commented:
@urso I have now addressed each of the comments from your latest review round and this PR is ready for another round of review.

For most of your comments (i.e. the ones to which I added a 👍), I made updates to the PR per your suggestions. But there are a few comments for which I requested clarification or furthered the discussion, e.g. [here](https://github.com/elastic/beats/pull/9260#discussion_r268355154), [here](https://github.com/elastic/beats/pull/9260#discussion_r268324987), [here](https://github.com/elastic/beats/pull/9260#discussion_r268361807) and [here](https://github.com/elastic/beats/pull/9260#discussion_r268356809).

Thanks!

ycombinator commented:
@urso Addressed your latest feedback, concretely: https://github.com/elastic/beats/pull/9260#discussion_r270182739, https://github.com/elastic/beats/pull/9260#discussion_r270178607, and https://github.com/elastic/beats/pull/9260#discussion_r270180130. Ready for review again, thanks!

ycombinator commented:
Got some feedback from @urso off-PR so mentioning them here:

> Can you add some more extra godoc to the package, explaining that it is for the ES reporter only that supports 2 different APIs please?

> `report.ReportingFormat` kind of stutters. With having them in a separate package how about changing the naming to `report.Format` ?

I will address both points in this PR shortly.

ycombinator commented:
jenkins, test this

ycombinator commented:
jenkins, test this

ycombinator commented:
jenkins, test this

ycombinator commented:
jenkins, test this

ycombinator commented:
jenkins, test this

cachedout commented:
🎉 🍾
